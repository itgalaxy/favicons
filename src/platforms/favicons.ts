import { FaviconHtmlElement } from "../index";
import { FaviconOptions, NamedIconOptions } from "../config/defaults";
import { transparentIcon, transparentIcons } from "../config/icons";
import { OptionalMixin, Platform, uniformIconOptions } from "./base";

const ICONS_OPTIONS: (NamedIconOptions & OptionalMixin)[] = [
  { name: "favicon.ico", ...transparentIcons(16, 24, 32, 48, 64) },
  { name: "favicon-16x16.png", ...transparentIcon(16) },
  { name: "favicon-32x32.png", ...transparentIcon(32) },
  { name: "favicon-48x48.png", ...transparentIcon(48) },
  { name: "favicon.svg", ...transparentIcon(1024), optional: true }, // arbitrary size. if more than one svg source is given, the closest to this size will be picked.
];

export class FaviconsPlatform extends Platform {
  constructor(options: FaviconOptions) {
    super(
      options,
      uniformIconOptions(options, options.icons.favicons, ICONS_OPTIONS),
    );
  }

  override async createHtml(): Promise<FaviconHtmlElement[]> {
    return this.iconOptions.map(({ name, ...options }) => {
      if (name.endsWith(".ico")) {
        // prettier-ignore
        return `<link rel="icon" type="image/x-icon" href="${this.cacheBusting(this.relative(name))}">`;
      } else if (name.endsWith(".svg")) {
        // prettier-ignore
        return `<link rel="icon" type="image/svg+xml" href="${this.cacheBusting(this.relative(name))}">`;
      }

      const { width, height } = options.sizes[0];

      // prettier-ignore
      return `<link rel="icon" type="image/png" sizes="${width}x${height}" href="${this.cacheBusting(this.relative(name))}">`;
    });
  }
}
