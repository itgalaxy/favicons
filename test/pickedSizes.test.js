import favicons from "../src";
import { logo_png } from "./util";

test("should generate favicons of a given size", async () => {
  expect.assertions(1);
  const result = await favicons(logo_png, {
    icons: {
      favicons: [
        {
          name: "favicon.ico",
          sizes: [
            { width: 48, height: 48 },
            { width: 64, height: 64 },
          ],
        },
        "favicon-48x48.png",
      ],
      android: false,
      appleIcon: false,
      appleStartup: false,
      windows: false,
      yandex: false,
    },
  });
  await expect(result).toMatchFaviconsSnapshot();
});
