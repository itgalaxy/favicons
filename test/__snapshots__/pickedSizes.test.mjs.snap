// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate favicons of a given size 4`] = `
{
  "files": [],
  "html": [
    "<link rel="icon" type="image/x-icon" href="/favicon.ico">",
    "<link rel="icon" type="image/png" sizes="48x48" href="/favicon-48x48.png">",
  ],
  "htmlTags": [
    FaviconHtmlTag {
      "attrs": {
        "href": "/favicon.ico",
        "rel": "icon",
        "type": "image/x-icon",
      },
      "tag": "link",
    },
    FaviconHtmlTag {
      "attrs": {
        "href": "/favicon-48x48.png",
        "rel": "icon",
        "sizes": "48x48",
        "type": "image/png",
      },
      "tag": "link",
    },
  ],
  "images": [
    {
      "contents": null,
      "name": "favicon.ico",
    },
    {
      "contents": null,
      "name": "favicon-48x48.png",
    },
  ],
}
`;
